We go over making the cnn from scratch using the functional and sequantial api

added batch normalization to it
https://www.youtube.com/watch?v=DtEq44FTPM4&ab_channel=CodeEmporium

https://www.youtube.com/watch?v=nUUqwaxLnWs&t=4s&ab_channel=DeepLearningAI

Batch Normilization Paper -> https://arxiv.org/abs/1502.03167

Batch Normilization and how it helps optimization:
https://arxiv.org/abs/1805.11604

Batch Normilization:

Increases speed of training:

* so if one variable has a small range, like height which will go from 0-2.5metres, and then another which is age that goes from 0 to 80, any small change in the height will make the updates go crazy, so you need a small learning rate
* Normalization can make mean 0 and standard deviation 1, so can use greater learning rate, to get to minimum faster

Decresed importance of weight initilization:
*becasue of the above stuf, the graph is alot smoother so you can sample your starting points between -1 and 1, also then youll reach the minimum


- also has a tiny regularization effect

In [2]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras.datasets import cifar10

In [3]:
#Load Data
(xtrain,ytrain),(xtest,ytest)=cifar10.load_data()

# convert the data to float and also normalize to between 1 and 0
xtrain= xtrain.astype("float32")/255.0
xtest=xtest.astype("float32")/255.0


170498071/170498071 [==============================] - 14s 0us/step


In [ ]:
xtrain.shape

(50000, 32, 32, 3)

In [4]:
# Create Model, no need to flatten data because you are using a CNN

model = keras.Sequential(
    [
    keras.Input(shape=(32,32,3)),
    layers.Conv2D(32,3,1,padding='same',activation='relu',input_shape=(32,3,3)),
    layers.MaxPooling2D(2),
    layers.Conv2D(64,3,1,padding='same',activation='relu'),
    layers.MaxPooling2D(2),
    layers.Conv2D(64,3,1,padding='same',activation='relu'),

    # now these layers have to go into the dense layers, ie. the fully connected part
    layers.Flatten(),
    layers.Dense(64,activation='relu'),
    layers.Dense(10)

    ]

)

model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 32, 32, 32)        896       
                                                                 
 max_pooling2d (MaxPooling2  (None, 16, 16, 32)        0         
 D)                                                              
                                                                 
 conv2d_1 (Conv2D)           (None, 16, 16, 64)        18496     
                                                                 
 max_pooling2d_1 (MaxPoolin  (None, 8, 8, 64)          0         
 g2D)                                                            
                                                                 
 conv2d_2 (Conv2D)           (None, 8, 8, 64)          36928     
                                                                 
 flatten (Flatten)           (None, 4096)              0

In [ ]:
# compile model

model.compile(
    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
    optimizer=keras.optimizers.Adam(lr=3e-4),
    metrics=["accuracy"]
)

model.fit(xtrain,ytrain,batch_size=128,epochs=10,verbose=2)
model.evaluate(xtest,ytest,batch_size=128,verbose=2)

Epoch 1/10
391/391 - 13s - loss: 1.5502 - accuracy: 0.4320 - 13s/epoch - 33ms/step
Epoch 2/10
391/391 - 1s - loss: 1.1643 - accuracy: 0.5887 - 1s/epoch - 4ms/step
Epoch 3/10
391/391 - 1s - loss: 0.9938 - accuracy: 0.6509 - 1s/epoch - 4ms/step
Epoch 4/10
391/391 - 1s - loss: 0.8724 - accuracy: 0.6958 - 1s/epoch - 4ms/step
Epoch 5/10
391/391 - 2s - loss: 0.7919 - accuracy: 0.7243 - 2s/epoch - 4ms/step
Epoch 6/10
391/391 - 1s - loss: 0.7299 - accuracy: 0.7463 - 1s/epoch - 4ms/step
Epoch 7/10
391/391 - 2s - loss: 0.6726 - accuracy: 0.7666 - 2s/epoch - 5ms/step
Epoch 8/10
391/391 - 2s - loss: 0.6297 - accuracy: 0.7803 - 2s/epoch - 4ms/step
Epoch 9/10
391/391 - 1s - loss: 0.5814 - accuracy: 0.7972 - 1s/epoch - 4ms/step
Epoch 10/10
391/391 - 1s - loss: 0.5380 - accuracy: 0.8125 - 1s/epoch - 4ms/step
79/79 - 0s - loss: 0.7820 - accuracy: 0.7398 - 375ms/epoch - 5ms/step


[0.782039999961853, 0.739799976348877]

WARNING:absl:`lr` is deprecated in Keras optimizer, please use `learning_rate` or use the legacy optimizer, e.g.,tf.keras.optimizers.legacy.Adam.
Epoch 1/10
391/391 - 13s - loss: 1.5502 - accuracy: 0.4320 - 13s/epoch - 33ms/step
Epoch 2/10
391/391 - 1s - loss: 1.1643 - accuracy: 0.5887 - 1s/epoch - 4ms/step
Epoch 3/10
391/391 - 1s - loss: 0.9938 - accuracy: 0.6509 - 1s/epoch - 4ms/step
Epoch 4/10
391/391 - 1s - loss: 0.8724 - accuracy: 0.6958 - 1s/epoch - 4ms/step
Epoch 5/10
391/391 - 2s - loss: 0.7919 - accuracy: 0.7243 - 2s/epoch - 4ms/step
Epoch 6/10
391/391 - 1s - loss: 0.7299 - accuracy: 0.7463 - 1s/epoch - 4ms/step
Epoch 7/10
391/391 - 2s - loss: 0.6726 - accuracy: 0.7666 - 2s/epoch - 5ms/step
Epoch 8/10
391/391 - 2s - loss: 0.6297 - accuracy: 0.7803 - 2s/epoch - 4ms/step
Epoch 9/10
391/391 - 1s - loss: 0.5814 - accuracy: 0.7972 - 1s/epoch - 4ms/step
Epoch 10/10
391/391 - 1s - loss: 0.5380 - accuracy: 0.8125 - 1s/epoch - 4ms/step
79/79 - 0s - loss: 0.7820 - accuracy: 0.7398 - 375ms/epoch - 5ms/step
[0.782039999961853, 0.739799976348877]


If you look athe above, then you can see the accuracy isnt good

this is because we are overfitting  - so you can add regularization to fix it

In [ ]:
# build CNN using functional api
inputs=keras.Input(shape=(32,32,3))
#you dont need to add activation below, becasue we are doing batch normilization, so we do activation after it
x=layers.Conv2D(32,3)(inputs)
x=layers.BatchNormalization()(x)
x=keras.activations.relu(x)
#if you dont specify anything, it defaults to a 2x2
x=layers.MaxPooling2D()(x)
x=layers.Conv2D(64,5,padding='same')(x)
x=layers.BatchNormalization()(x)
x=keras.activations.relu(x)
x=layers.Conv2D(128,3)(x)
x=layers.BatchNormalization()(x)
x=keras.activations.relu(x)
x=layers.Flatten()(x)
x=layers.Dense(64,activation='relu')(x)
outputs=layers.Dense(10,activation='softmax')(x)
model=keras.Model(inputs=inputs,outputs=outputs)

model.compile(
    loss=keras.losses.SparseCategoricalCrossentropy(),
    optimizer=keras.optimizers.Adam(lr=3e-4),
    metrics=["accuracy"]
)

model.fit(xtrain,ytrain,batch_size=64,epochs=10,verbose=2)
model.evaluate(xtest,ytest,batch_size=64,verbose=2)


Epoch 1/10
782/782 - 7s - loss: 1.5502 - accuracy: 0.4435 - 7s/epoch - 9ms/step
Epoch 2/10
782/782 - 3s - loss: 1.0682 - accuracy: 0.6157 - 3s/epoch - 4ms/step
Epoch 3/10
782/782 - 4s - loss: 0.8901 - accuracy: 0.6808 - 4s/epoch - 5ms/step
Epoch 4/10
782/782 - 4s - loss: 0.7796 - accuracy: 0.7249 - 4s/epoch - 5ms/step
Epoch 5/10
782/782 - 3s - loss: 0.6962 - accuracy: 0.7525 - 3s/epoch - 4ms/step
Epoch 6/10
782/782 - 3s - loss: 0.6190 - accuracy: 0.7810 - 3s/epoch - 4ms/step
Epoch 7/10
782/782 - 4s - loss: 0.5487 - accuracy: 0.8053 - 4s/epoch - 5ms/step
Epoch 8/10
782/782 - 3s - loss: 0.4879 - accuracy: 0.8284 - 3s/epoch - 4ms/step
Epoch 9/10
782/782 - 3s - loss: 0.4248 - accuracy: 0.8498 - 3s/epoch - 4ms/step
Epoch 10/10
782/782 - 4s - loss: 0.3703 - accuracy: 0.8688 - 4s/epoch - 5ms/step
157/157 - 1s - loss: 0.9911 - accuracy: 0.7077 - 533ms/epoch - 3ms/step


[0.9911022186279297, 0.7077000141143799]

In [7]:
#can add dropout, data augmentation and L2 regularization to stop overfitting, or early stopage
#https://www.youtube.com/playlist?list=PLkDaE6sCZn6Hn0vK8co82zjQtt3T2Nkqc

####################
from tensorflow.keras import regularizers

####################
# build CNN using functional api
inputs=keras.Input(shape=(32,32,3))
#you dont need to add activation below, becasue we are doing batch normilization, so we do activation after it
x=layers.Conv2D(32,3,padding='same',kernel_regularizer=regularizers.l2(0.01))(inputs)
x=layers.BatchNormalization()(x)
x=keras.activations.relu(x)
#if you dont specify anything, it defaults to a 2x2
x=layers.MaxPooling2D()(x)
x=layers.Conv2D(64,3,padding='same',kernel_regularizer=regularizers.l2(0.01))(x)
x=layers.BatchNormalization()(x)
x=keras.activations.relu(x)
x=layers.Conv2D(128,3,padding='same',kernel_regularizer=regularizers.l2(0.01))(x)
x=layers.BatchNormalization()(x)
x=keras.activations.relu(x)
x=layers.Flatten()(x)
x=layers.Dense(64,activation='relu')(x)
x=layers.Dropout(0.5)(x)
outputs=layers.Dense(10,activation='softmax')(x)
model=keras.Model(inputs=inputs,outputs=outputs)

model.compile(
    loss=keras.losses.SparseCategoricalCrossentropy(),
    optimizer=keras.optimizers.Adam(lr=3e-4),
    metrics=["accuracy"]
)

model.fit(xtrain,ytrain,batch_size=64,epochs=150,verbose=2)
model.evaluate(xtest,ytest,batch_size=64,verbose=2)


Epoch 1/150
782/782 - 17s - loss: 2.5345 - accuracy: 0.1309 - 17s/epoch - 22ms/step
Epoch 2/150
782/782 - 5s - loss: 2.2160 - accuracy: 0.1396 - 5s/epoch - 6ms/step
Epoch 3/150
782/782 - 5s - loss: 2.2016 - accuracy: 0.1425 - 5s/epoch - 6ms/step
Epoch 4/150
782/782 - 5s - loss: 2.2053 - accuracy: 0.1409 - 5s/epoch - 6ms/step
Epoch 5/150
782/782 - 5s - loss: 2.1982 - accuracy: 0.1423 - 5s/epoch - 6ms/step
Epoch 6/150
782/782 - 5s - loss: 2.1948 - accuracy: 0.1460 - 5s/epoch - 6ms/step
Epoch 7/150
782/782 - 5s - loss: 2.1899 - accuracy: 0.1463 - 5s/epoch - 6ms/step
Epoch 8/150
782/782 - 5s - loss: 2.1909 - accuracy: 0.1455 - 5s/epoch - 6ms/step
Epoch 9/150
782/782 - 5s - loss: 2.1796 - accuracy: 0.1470 - 5s/epoch - 6ms/step
Epoch 10/150
782/782 - 5s - loss: 2.1838 - accuracy: 0.1462 - 5s/epoch - 6ms/step
Epoch 11/150
782/782 - 5s - loss: 2.1767 - accuracy: 0.1479 - 5s/epoch - 6ms/step
Epoch 12/150
782/782 - 5s - loss: 2.1689 - accuracy: 0.1502 - 5s/epoch - 7ms/step
Epoch 13/150
782/782 -

[1.6178300380706787, 0.47690001130104065]